Cases advanced search

northwestwitch opened the pull request:
This PR adds a functionality or fixes a bug.
- Closes #4704

<img width="1450" alt="image" src="https://github.com/user-attachments/assets/e13d73ba-f59b-437f-8302-c487e8f453a8">

### Toggle advanced search:

<img width="955" alt="image" src="https://github.com/user-attachments/assets/95224caf-4777-4afe-acef-a201df5872af">


<details>
<summary>Testing on cg-vm1 server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. Make sure the PR is pushed and available on [Docker Hub](https://hub.docker.com/repository/docker/clinicalgenomics/scout-server-stage)
1. Fist book your testing time using the Pax software available at [https://pax.scilifelab.se/](https://pax.scilifelab.se). The resource you are going to call dibs on is `scout-stage` and the server is `cg-vm1`.
1. `ssh <USER.NAME>@cg-vm1.scilifelab.se`
1. `sudo -iu hiseq.clinical`
1. `ssh localhost`
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `podman ps`
1. Stop the service with current deployed branch: `systemctl --user stop scout.target`
1. Start the scout service with the branch to test: `systemctl --user start scout@<this_branch>`
1. Make sure the branch is deployed: `systemctl --user status scout.target`
1. After testing is done, repeat procedure at [https://pax.scilifelab.se/](https://pax.scilifelab.se), which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>

<details>
<summary>Testing on hasta server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. `ssh <USER.NAME>@hasta.scilifelab.se`
1. Book your testing time using the Pax software. `us; paxa -u <user> -s hasta -r scout-stage`. You can also use the WSGI Pax app available at [https://pax.scilifelab.se/](https://pax.scilifelab.se).
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `conda activate S_scout; pip freeze | grep scout-browser`
1. Deploy the branch to test: `bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_scout -t scout -b <this_branch>`
1. Make sure the branch is deployed: `us; scout --version`
1. After testing is done, repeat the `paxa` procedure, which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>


**How to test**:
1. Deploy on stage and try some advanced search

**Expected outcome**:
The functionality should be working
Take a screenshot and attach or copy/paste the output.

**Review:**
- [ ] code approved by
- [ ] tests executed by

northwestwitch commented on scout/adapter/mongo/case.py:
```diff
@@ -102,24 +103,20 @@ def _set_similar_phenotype_query(
-        order = []
         for i in similar_cases:
             similar_case_ids.append(i[0])
-            order.append(i[1])
```
This is weird and I don't know how we were thinking when we wrote the function like this. 

Testing locally: `similar_cases` looks like this:

<img width="178" alt="image" src="https://github.com/user-attachments/assets/5e617fc2-8d78-4d97-a266-1a576c1c5b17">

So `order` will be:  [1.0, 1.0, 1.0]

This value is returned by the function.

northwestwitch commented on scout/adapter/mongo/case.py:
```diff
@@ -494,21 +487,20 @@ def _conditional_set_query_value(query, condition, set_key, set_value):
         if yield_query:
             return query
 
-        if order:
```
Then order is used here, just to say if the array has values use the following projection:

'analysis_date': 1, 'assignees': 1, 'beacon': 1, 'case_id': 1, 'display_name': 1, 'individuals': 1, 'is_rerun': 1, 'is_research': 1, 'mme_submission': 1, 'owner': 1, 'panels': 1, 'status': 1, 'track': 1, 'vcf_files': 1}

I don't think the order thing adds anything by itself in here?

dnil replied:
Hm, I'm probably slow today, but at least I'm looking.. 😆 It returns them in the order given, without sorting, and the intention appears to be that they should appear in the same order as the "case_id: $in" operation gave them previously. Are you saying this doesn't work anyway, or that the returned similarity values are the same so it doesn't work? 😊

northwestwitch commented on scout/adapter/mongo/case.py:
```diff
@@ -102,24 +103,20 @@ def _set_similar_phenotype_query(
             similar_cases = self.cases_by_phenotype(hpo_terms, institute_id, case_obj["_id"])
         else:  # similar HPO terms
-            hpo_terms = list(query_term.replace(" ", "").split(","))
+            hpo_terms = list(query_value.replace(" ", "").split(","))
```
Just renamed he variable to be more clear

dnil replied:
💯

dnil replied:
Query term was more logical when the term was in there, and not in a separate field I guess.. 😁

northwestwitch commented:
@dnil unless you have started reviewing let's not include this PR in the next release I think

dnil commented:
> @dnil unless you have started reviewing, let's not include this PR in the next release I think

Sounds like a plan! Will be really good, but may take a little more iterating. Query changes and all that.. 😅

dnil commented on scout/adapter/mongo/case.py:
```diff
@@ -201,94 +193,107 @@ def _set_diagnosis_query(self, query: Dict[str, Any], query_term: str):
-            self._set_diagnosis_query(query, query_term)
+    def update_case_query_or_options(self, query: dict, or_options: List[dict]):
+        """Populates the available options in the query $and field"""
+        if or_options and query.get("$or"):
```
i
```suggestion
        if not or_options:
            return
       if query.get("$or"):
```

dnil commented on scout/adapter/mongo/case.py:
```diff
@@ -201,94 +193,107 @@ def _set_diagnosis_query(self, query: Dict[str, Any], query_term: str):
+        """Populates the available options in the query $and field"""
+        if or_options and query.get("$or"):
+            query["$or"] += or_options
+        elif or_options:
```
```suggestion
```

northwestwitch replied:
I fixed it like this:

```
        if not or_options:
            return
        query.setdefault("$or", []).extend(or_options)
```

dnil commented on scout/adapter/mongo/case.py:
```diff
@@ -201,94 +193,107 @@ def _set_diagnosis_query(self, query: Dict[str, Any], query_term: str):
+                    )
 
-    def _update_case_id_query(self, query, id_list):
+    def update_case_id_query(self, query, id_list):
```
Huh, is this really supposed to be imported outside the module now? Or am I being archaic with PEP-8?
![Screenshot 2024-08-28 at 08 12 57](https://github.com/user-attachments/assets/49287e0d-960b-48b1-8316-fcb18d799606)

northwestwitch replied:
Mmm no, perhaps I intended to import it from some other place at one point, but I didn't do it. I'll revert!

dnil commented on scout/adapter/mongo/case.py:
```diff
@@ -201,94 +193,107 @@ def _set_diagnosis_query(self, query: Dict[str, Any], query_term: str):
-            order = self._set_similar_phenotype_query(
-                query, query_field, query_term, owner or collaborator
-            )
+    def populate_case_query(
```
Also it's not quite working? I tried filling in a couple of search fields: anything genes related gets a 
<img width="962" alt="Screenshot 2024-08-28 at 08 27 39" src="https://github.com/user-attachments/assets/7a8021c4-2715-48a1-9469-4736a41813d8">

northwestwitch replied:
ChatGPT! 😆 I modified and removed the lambdas, but it wasn't so hard to follow I think. I mean, historically I'm the one no getting the lambda thing, so I reasoned that if I understood then everybody could. But sure, without lambdas is more readable!

dnil replied:
I do like lambdas for functional stuff, like in sorting or for more map-reduce like operation, but mixed in for some functions in a large list, I think this is cleaner.

dnil commented on scout/adapter/mongo/case.py:
```diff
@@ -521,7 +522,7 @@ def rna_cases(self, owner):
             list of case _ids
         """
-        EXISTS_NOT_NULL = {"$exists": True, "$nin": [None, ""]}
+        EXISTS_NOT_NULL = {EXISTS: True, "$nin": [None, ""]}
```
I dunno, this looks like pleasing some silly code evaluation? It's not like mongo are going to change that function anytime soon?! And if so we could search-replace it just as easy. The rule is for different situations, like natural language to be translated, or more complex ones that could go wrong. "EXISTS_NOT_NULL" is good, and e.g. `EXISTS = {"$exists": True}` or whatnot would be ok in my book!

northwestwitch replied:
Yes, of course it's pleasing that code checker, I would never think of such replacement myself. I like your suggestion but if you check in the code you'll find that the "exists" is used in slightly different situations, for instance:

- set_value={EXISTS: False}
- set_value={EXISTS: True, "$ne": []} # there are a bunch of these
- set_value={EXISTS: True, "$eq": True},
- EXISTS_NOT_NULL = {EXISTS: True, "$nin": [None, ""]}

That's why I decided to replace only "$exists" and not the whole dict. This way it is more easy to read and doesn't trigger the code checker. I'd like to leave it like this if it's not a problem for you

northwestwitch replied:
I feel that if I replace the cases above with 5 different constants then it will be much less readable. If you don't like the EXISTS = "$exists" replacement I can simply revert to how the code was before and we ignore the code checker warning instead

dnil replied:
Whatever! It's just a constant string regardless, and the only reason to change should be to make it more readable and maintainable. Since there is a "EXISTS_NOT_NULL" already, adding the others as similar full constants would make sense to me, but its no big deal.

dnil commented on scout/server/blueprints/institutes/controllers.py:
```diff
@@ -10,13 +10,7 @@
-    ID_PROJECTION,
-    PHENOTYPE_GROUPS,
-)
+from scout.constants import CASE_STATUSES, DATE_DAY_FORMATTER, ID_PROJECTION, PHENOTYPE_GROUPS
```
👍

dnil commented on scout/server/blueprints/institutes/controllers.py:
```diff
@@ -435,20 +429,14 @@ def cases(store, request, institute_id):
     limit = int(request.args.get("search_limit")) if request.args.get("search_limit") else 100
-    data["form"] = populate_case_filter_form(request.args)
+
+    data["form"] = CaseFilterForm(request.form)
```
👍

dnil commented on CHANGELOG.md:
```diff
@@ -22,6 +22,7 @@ About changelog [here](https://keepachangelog.com/en/1.0.0/)
 
 ## [4.87]
 ### Added
+- Advanced cases search to narrow down results using more than one search parameter
```
Careful with the changelog here! 😊

northwestwitch commented:
I'm ready for another review with this one. And I guess we can include it in the next minor, so I
'm not going to merge in the upcoming patch!

dnil commented on tests/adapter/mongo/test_case_handling.py:
```diff
@@ -209,12 +248,14 @@ def test_get_research_case(real_adapter, case_obj, institute_obj):
     assert adapter.case_collection.find_one()
 
-    # THEN searching for reasearch cases should return one case
+    # THEN searching for research cases should return one case
```
👌

dnil reviewed (commented):
Definitely getting there! Works for me, but I have a couple of comments regarding the handling and search feeling during testing, and also some small comment to the code. It's still a little bit back and forth with local functions and "suggested private" class methods as well, but good enough methinks, and a less strenuous read than the initial AI code.
- It can be a bit annoying that the advanced query accordion folds back in after search if you wanted to test a few different things.
- This is compounded by values staying populated - as they should. It is too easy to get confused why cases don't show up and realise you at one point typed something in an "advanced" field previously and forgot about it. Perhaps force the advanced tab open if anything is filled in, or indicate somewhere next to the case name search field that advanced options are in play?
- A "clear search" option would be useful with this many fields, but could be added later.

northwestwitch commented on scout/server/blueprints/institutes/templates/overview/cases.html:
```diff
@@ -22,49 +22,116 @@
-      </div>
-      <div class="col-3">
-          <label class="control-label" for="search_term">Search cases</label>
-          <input class="form-control" id="search_term" name="search_term" value="{{form.search_term.data if form.search_term.data}}" placeholder="example:18201" type="text" pattern="[^\\\<\>\?\!\=\/]*" title="Characters \<>?!=/ are not allowed">
```
Ah right, good catch!, I'll re-add it!

northwestwitch replied:
Fixed!

northwestwitch commented:
> Definitely getting there! Works for me, but I have a couple of comments regarding the handling and search feeling during testing, and also some small comment to the code. It's still a little bit back and forth with local functions and "suggested private" class methods as well, but good enough methinks, and a less strenuous read than the initial AI code.
> 
> * It can be a bit annoying that the advanced query accordion folds back in after search if you wanted to test a few different things.
> * This is compounded by values staying populated - as they should. It is too easy to get confused why cases don't show up and realise you at one point typed something in an "advanced" field previously and forgot about it. Perhaps force the advanced tab open if anything is filled in, or indicate somewhere next to the case name search field that advanced options are in play?
> * A "clear search" option would be useful with this many fields, but could be added later.

Good suggestions, I fixed it as you said:

<img width="899" alt="image" src="https://github.com/user-attachments/assets/9e74106a-10dd-45c1-ad16-c220bb080aba">

- filters stays open when you perform an advanced search
- At the bottom it shows the keys/values used in the search
- Added a clear search button

dnil commented on scout/server/blueprints/institutes/templates/overview/cases.html:
```diff
@@ -27,7 +27,7 @@
         <div class="col-md-3 mb-3">
             {{ form.case.label(class="form-label") }}
-            {{ form.case(class="form-control", placeholder="example:18201") }}
+            {{ form.case(class="form-control", placeholder="example:18201", pattern="[^\\\<\>\?\!\=\/]*", title="Characters \<>?!=/ are not allowed") }}
```
Hm, I wonder if we needed it for the other query fields as well?

northwestwitch replied:
Fix, thanks. Better to be safe than sorry!

northwestwitch commented:
Searching by similar case is actually broken, like you get an exception. I'll push a fix (not the fix to #4839, but a simple one)

dnil commented:
Ahhh, I was about to say it works on stage, but just tested on prod. It no longer (if ever) works since the empty match can become `None` instead of `[]`. Good catch.

northwestwitch commented:
> Searching by similar case is actually broken, like you get an exception. I'll push a fix (not the fix to #4839, but a simple one)

It happens when you look for cases similar to a case without phenotype. An existing bug I presume

northwestwitch commented:
This one should be ready also @dnil

dnil commented:
Hm, it still removes the old, arguably anyway partly broken order. Let's discuss that one - I would prefer to fix or remove it as part of #4882, but either fully or not here. I'm open to arguments though!

northwestwitch commented:
> Hm, it still removes the old, arguably anyway partly broken order. Let's discuss that one - I would prefer to fix or remove it as part of #4839 , but either fully or not here. I'm open to arguments though!

I don't want to fix that order thing on this PR. It's already quite long. I'll attempt a fix in a new PR

northwestwitch commented on scout/adapter/mongo/case.py:
```diff
@@ -108,13 +102,21 @@ def _set_similar_phenotype_query(
+        similar_case_ids = [
+            similar_case[0] for similar_case in similar_cases
+        ]  # These are sorted by the most similar to the least
+        self._update_case_id_query(
```
This advanced case search is quite delicate, because if you have other params that will affect the content of query["_id"]["$in"] you can mess up the order of the similar cases

northwestwitch commented on scout/adapter/mongo/case.py:
```diff
@@ -197,86 +194,149 @@ def _set_diagnosis_query(self, query: Dict[str, Any], query_term: str):
+                "pinned",
+                "causative",
+                "user",
+                "similar_case",  # In order to be able to sort results by phenotype similarity, keep this at the bottom
```
I have moved these 2 at the bottom so whatever happens before setting the query _id values, the order by similarity will be maintained (hopefully)

northwestwitch commented:
I've included also the "search by case similarity" fix now. Tested with the combo search on cust003

{'cohort': 'hello', 'similar_case': 'holydassie', 'search_limit': 100}:

And it works, returning this case -> cout-stage.scilifelab.se/cust003/16062

dnil reviewed (approved):
Super, thank you, and especially for lifting the order fix part to the other PR! As before, this will be very useful, EXIST or not. 😁 Let's also make an issue for moving the dashboard search to this format, but no need to do in this PR. 😊

northwestwitch commented:
> Let's also make an issue for moving the dashboard search to this format, but no need to do in this PR

Exactly, and if we do that the code will look clearer!
